hermeto/core/package_managers/gomod.py

```diff
@@ -964,6 +967,83 @@ def _setup_go_toolchain(go_mod_file: RootedPath) -> Go:
+
+    go_version_str, toolchain_version_str = _get_gomod_version(go_mod_file)
+    log.debug(
+        go_mod_version_msg,
+        go_version_str if go_version_str else "-",
+        toolchain_version_str if toolchain_version_str else "-",
+    )
+
+    if not go_version_str:
+        # Go added the 'go' directive to go.mod in 1.12 [1]. If missing, 1.16 is assumed [2].
+        # For our version comparison purposes we set the version explicitly to 1.20 if missing.
+        # [1] https://go.dev/doc/go1.12#modules
+        # [2] https://go.dev/ref/mod#go-mod-file-go
+        go_version_str = "1.20"
+        log.debug("Could not parse Go version from go.mod, using %s as fallback", go_version_str)
+
+    if not toolchain_version_str:
+        toolchain_version_str = go_version_str
+
+    go_mod_version = GoVersion(go_version_str)
+    go_mod_toolchain_version = GoVersion(toolchain_version_str)
+
+    if go_mod_version >= go_mod_toolchain_version:
+        target_version = go_mod_version
+    else:
+        target_version = go_mod_toolchain_version
+
+    if target_version.language > go_max_version.language:
+        raise PackageManagerError(
+            f"Required/recommended Go toolchain version '{target_version}' is not supported yet.",
+            solution=(
+                "Please lower your required/recommended Go version and retry the request. "
+                "You may also want to open a feature request on adding support for this version."
+            ),
+        )
+
+    # If we cannot find a matching toolchain, we'll try to fallback to a 1.21 one
+    matching_toolchains = filter(lambda t: t.version >= target_version, installed_toolchains)
+    try:
+        go = min(matching_toolchains)
+        log.debug("Using Go toolchain version '%s'", go.version)
+    except ValueError:
+        try:
+            go = max(filter(lambda t: t.version >= GoVersion("1.21"), installed_toolchains))
+            log.debug("Best matching Go toolchain version: '%s'", go.version)
+            log.debug("Will use Go toolchain version '%s' [via GOTOOLCHAIN=auto]", target_version)
+        except ValueError:
+            # This is a long shot - we couldn't find a matching toolchain, we don't have a toolchain
+            # that can do GOTOOLCHAIN=auto, so we pick any installed toolchain (we know we have
+            # any) and use it to download a new full-blown SDK for the target version
+            log.debug("Installing Go toolchain version '%s'", target_version)
+            release_str = f"go{str(target_version)}"
+            try:
+                work_toolchain = next(iter(installed_toolchains))
+                go = Go.from_missing_toolchain(work_toolchain.binary, release=release_str)
+                log.debug("Using Go toolchain version '%s'", go.version)
+            except StopIteration:
+                # should not happen, we test this in fetch_gomod_source, this is only for unit tests
```

Comment:
Why would we want to test this behavior if we know that it is impossible?

Reply:
Strictly speaking from unit testing perspective that context is irrelevant. I'll drop that particular test combo, I never liked this exception handling myself anyway.